### DISCUSSION TITLE
How does clangd figure out what flags and include paths to use when analyzing a header file?

### DISCUSSION THREAD

**Discussion by itavero** [clangd/clangd#2115] [Q&A]
I was wondering if someone can explain to me how clangd determines which compiler flags and include paths to use when analyzing a header file, and how to determine if it's a C or C++ file.
I did find a [short section on it in the documentation](https://clangd.llvm.org/design/compile-commands#commands-for-header-files), but it didn't really help me in explaining what is happening in my code base.

The reason I'm asking is that we have several C++ header files (with a `.h` extension) that cause generate some warnings when analyzed on their own, amongst others:
* Certain system include paths are not found (for instance `#include <cstdint>`)
* It can't find certain header files for which we have put a dependency in place in our CMake build system.
* It gives warnings that make me think it does not know how to interpret the header file.

For the last point, I have a file with the following contents (I've stripped the comments):
```cpp
#pragma once

#include <cstdint>

enum class TaskId {
  kMainTaskId = 1,
  kTaskIdMax
};
constexpr uint8_t kNumberOfTasks = static_cast<uint8_t>(TaskId::kTaskIdMax) - 1u;
```

On which I get the following "squiggly" lines (a.k.a. warnings) in my IDE:

* On the `#include <cstdint>`:
  * > Included header cstdint is not used directly
  * > In included file: This file requires compiler and library support for the ISO C++ 2011 standard. This support must be enabled with the -std=c++11 or -std=gnu++11 compiler options.
* on the `enum class TaskId {` line:
  * > Tentative definition has type 'enum class' that is never completed `clang(tentative_def_incomplete_type)`
tasks_config.h(10, 6): Forward declaration of 'enum class'
  * > Invalid case style for variable 'TaskId' (fix available) `clang-tidy(readability-identifier-naming)`
* On the `kNumberOfTasks` line:
  * > Unknown type name 'constexpr' `clang(unknown_typename)`
  * > Expected ';' after top level declarator (fix available) `clang(invalid_token_after_toplevel_declarator)`

Interesting thing is that it does not issue these warnings for every header file (even if they have similar `#include` statements or keywords being used). Which makes me think that clangd might be doing something smart, which I wish to understand.

The `.clangd` config I use at the moment (partly "generated" by input from CMake) looks as follows and the query driver is set to a pattern that matches all `gcc` and `g++` compilers:
```yml
CompileFlags:
  CompilationDatabase: D:/gitrepo/myproject/build/armgcc-10-debug
  Add: []
  Remove: []
---
If:
  PathMatch: .*\.h
CompileFlags:
  Compiler: "C:/cc/Arm/10.3_2021.10/bin/arm-none-eabi-g++.exe"
  Add:
    [
      -xc++,
      -mcpu=cortex-m33,
      -mthumb,
      -mfloat-abi=hard,
      -mfpu=fpv5-sp-d16,
      -mcmse,
      -ffunction-sections,
      -fdata-sections,
      -Wall,
      -Wextra,
      -Wimplicit-fallthrough,
      -Wfloat-equal,
      -Wformat=2,
      -Wmissing-include-dirs,
      -Wshadow,
      -Wswitch-default,
      -Wlogical-op,
      -Wundef,
      -Wno-unused-but-set-parameter,
      -Wno-unused-parameter,
      -Wno-narrowing,
      -fno-common,
      -ffreestanding,
      -fno-builtin,
      -fdata-sections,
      -ffunction-sections,
      -Wold-style-cast,
      -Wsuggest-override,
      -Wredundant-tags,
      -Woverloaded-virtual,
      -fstrict-enums,
      -fno-rtti,
      -fno-exceptions,
      -include D:/gitrepo/myproject/build/armgcc-10-debug/compdef_arm-none-eabi-g++.h,
    ]
```

In VS Code, if I look in the logs generated by clangd, I also see that for some header files it states:
> Indexing c++17 standard library in the context of d:\gitrepo\myproject\hw_drivers\led_bicolor\include\led_bicolor.h

For others it states:
> Indexing c99 standard library in the context of d:\gitrepo\myproject\bsp\h01\include\tasks_config.h

This second file is an example of a C++ header that gets interpreted incorrectly. VS Code does show _C++_ in the right bottom as the file type.

As a side note, we do have several C source files (in other targets) in our project as well. However, the `tasks_config.h` file mentioned here, is only included by `.cpp` files (and in general we try to write our C header files as such that the C++ compiler can also deal with them correctly).

Hopefully someone can shed some light on the innerworkings of clangd related to this, so I can somehow convince it to process all the header files correctly.

**Comment by HighCommander4** [comment]
> Interesting thing is that it does not issue these warnings for every header file (even if they have similar `#include` statements or keywords being used). Which makes me think that clangd might be doing something smart, which I wish to understand.

You can read some details about this at https://clangd.llvm.org/faq#what-can-i-do-if-clangd-chooses-the-wrong-source-file-to-infer-commands-for-a-header and the sections linked from there.

Does your project have a `compile_commands.json` file? If not, and the only source of build flags is your `.clangd` file, then based on the diagnostics you're getting, it sounds like perhaps the issue is that the list of flags in `.clangd` is missing `-std=c++11`?

In any case, you're welcome to share [clangd logs](https://clangd.llvm.org/troubleshooting.html#gathering-logs) for help getting a more specific diagnosis of your issue.

**Reply by itavero** [reply]
Sorry, I failed to mention we indeed have a compilation database from CMake/Ninja available and are using it.

I'll have a look at the section you mentioned.

**Reply by itavero** [reply]
Unfortunately `compdb` doesn't work for my project (our compilation database lists sources from different drives on Windows and the tool doesn't like that at the moment), but I can work on the idea of extending the compilation database with the header files as well. CMake has all the information for this available.

After another look at the `.clangd` file I have CMake create, I also noticed that the language standard is not specified there yet. Might already fix some of the behavior if I change that.

**Comment by bsbernd** [comment]
Interesting discussion, I have the same issue with libfuse - include/fuse_common.h has
```
#if !defined(FUSE_H_) && !defined(FUSE_LOWLEVEL_H_)
#error "Never include <fuse_common.h> directly; use <fuse.h> or <fuse_lowlevel.h> instead."
#endif
```

and

```
#if !defined(FUSE_USE_VERSION) || FUSE_USE_VERSION < 30
#  error only API version 30 or greater is supported
#endif
```

and it always triggers these as errors for me. I'm going to test `CompDB`.

**Reply by HighCommander4** [reply]
That sounds like a different issue, see https://github.com/clangd/clangd/issues/45
